zkvm/src/signature/mod.rs

```diff
@@ -16,11 +16,59 @@ mod multikey;
 mod musig;
 mod signer;
 
-/// Verification key (aka "pubkey") is a wrapper type around a Ristretto point
+/// Verification key (aka "pubkey") is a wrapper type around a Ristretto point and it's compressed variation
 /// that lets the verifier to check the signature.
 #[derive(Copy, Clone, PartialEq, Eq, Debug)]
-pub struct VerificationKey(pub CompressedRistretto);
+pub struct VerificationKey {
+    point: RistrettoPoint,
+    // Available only if instantiated using `with_compressed`
+    // or if caller directly calls a method that returns a compressed point
+    compressed: Option<CompressedRistretto>,
+}
 
+impl VerificationKey {
+    /// Creates new key from a compressed form,remembers the compressed point.
+    pub fn with_compressed(p: CompressedRistretto) -> Option<Self> {
+        Some(VerificationKey {
+            point: p.decompress()?,
+            compressed: Some(p),
+        })
+    }
+    /// Creates a new key from a ristretto point. Does not eagerly compress.
+    pub fn with_decompressed(p: RistrettoPoint) -> Self {
+        VerificationKey {
+            point: p,
+            compressed: Some(p.compress()),
+        }
+    }
+    /// Compresses the point if needed (no-op if `self.compressed == Some`)
+    pub fn compress_in_place(&mut self) {
+        if self.compressed.is_none() {
+            self.compressed = Some(self.point.compress())
+        }
+    }
+    /// Converts the Verification key to a point
+    pub fn to_point(self) -> RistrettoPoint {
+        self.point
+    }
+
+    /// Converts the Verification key to a point and returns a reference to it
+    pub fn as_point(&self) -> &RistrettoPoint {
+        &self.point
+    }
+
+    /// Converts Verification key to a compressed point.
+    /// If the compressed form is not available, the compression is completed
+    pub fn to_compressed_point(&self) -> CompressedRistretto {
```

Comment:
Instead of to/as compressed/decompressed point we can impl `Deref<>` and `Into<>` for each of these types.